Баландин Игорь

balandin228 opened the pull request:
@xackill

balandin228 committed:
work on courses

balandin228 committed:
add tagCloud and tests

balandin228 committed:
refactor client architecture

balandin228 committed:
creating image with Result<T> logic

balandin228 committed:
major changes

group cs files
change Result<T> logic
change IAction
change actions logic
add Result tests
refactor some tests

balandin228 committed:
add help command + make settings is IClonable

balandin228 committed:
refactor IAction name

balandin228 committed:
add parsing sentence

xackill commented:
#### Общие замечания к тексту ниже
1. Любой из пунктов стоит обдумывать и холиварить, если не согласен. Мои слова, как и слова любого человека, **не могут быть приняты как истина в последней инстанции**.
2. Но все-таки хочется, чтобы **ни одно замечание** при этом **не было потеряно**: я считаю каждое из них полезным, но вот повторять написанный текст считаю тратой времени.

xackill commented:
#### Общий комментарий к коммитам
Сейчас есть какой-то набор изменений, однако он не прибран, как следствие его невозможно изучать раздельно. А значит для валидации решения необходимо скачать все решение и изучить его целиком. Можно заметить, насколько выросла сложность этого действия от TagCloud1 к TagCloud3.

Многие используют такую практику ревью ("посмотрел разницу между первым и последним коммитом"), однако у меня она выходит недостаточно качественной: не получается валидировать построчно (нельзя держать построчно такой контекст в голове). Поэтому я валидирую все покоммитно, а в коммитах — построчно.

Если при разработке полагать, что процесс ревью будет покоммитным, то приходится:
1. Делать коммиты гранулярным, т.е. 1 коммит решает 1 простую логически-законченную задачу (например, можно было так: "Добавление проекта в солюшен", "Перенос CircularCloudLayouter из TagCloud1", "Перенос Client из TagCloud2", "Добавление паттерна Result<T> для ...");
2. Эти гранулярные коммиты в идеале не должны переписывать друг друга (менять одну и ту же строчку несколько раз), а так же содержать "лишних" изменений (логически относящиеся к другим коммитам);
3. В каждом коммите солюшен билдится;
4. Цепочка коммитов должна понятным образом решать задачу.

Такой подход усложняет разработку (приходится еще и прибирать коммиты, переразбивать их перед пушем на сервер), однако очень ускоряет процесс ревью, повышает его качество, а самое важное, с твоим кодом потом сможет разобраться кто угодно просто по репозиторию.

Тут можно заметить, что это не самые распространенные паттерны разработки и ревью. Такой подход заставляет разработчика (тебя) тратить больше времени (может поэтому он и не очень популярен), однако я считаю, что он сильно экономит время на всех последующих стадиях (других людей), собственно поэтому я его в основном и практикую, в том числе и в ревью.

Предлагаю для правок попробовать именно такой подход.

xackill commented:
#### Общий комментарий к решению
Сейчас паттерн `Result<T>` применен точечно, нет какой-то общей схемы. В результате, код сложно валидировать (может тут быть исключение? или нет?). Для поддерживаемого решения это не есть гуд, поэтому для практики предлагаю поступить так:
1. В явном виде применить паттерн в общем цикле (сейчас там что-то гибридное, я к этому вернусь в отдельных комментариях);
2. В явном виде применить паттерн для всех классов, реализующих `IConsoleAction`;
3. Порефакторить применение паттерна для мест, где сейчас он усложняет чтение, а не упрощает (на этом я тоже остановлюсь подробнее).

- Что я подразумеваю под "в явном виде применить паттерн": это `chaining Result<T>`. Сам паттерн `Result<T>` исполнен так, что обрабатывает ошибки при помощи конструкции `Try...catch`, которая гарантирует поимку (и обработку) всех исключений, в то время как текущий подход (с точечными `if`-ами) не настолько надежен, в том числе для расширения. Например, `Console.ReadLine()` может выбрасывать исключение. Но точечными if-ами оно не обрабатывается (да и накладно это), поэтому его верхнеуровнево (в методе, где вызывается) необходимо обернуть в `Result<T>`.

- Отдельно замечу, что сейчас в решении присутствует достаточно много неиспользуемого кода; его стоит поудалять, почистить решение.

xackill commented on TagCloud/Client.cs:
```diff
@@ -25,123 +25,36 @@ public Client(IAction[] actions, IPaletteNamesFactory paletteNamesFactory)
-        public void Start(ICloudVisualization visualization)
+        public void Start()
         {
             availablePaletteNames = paletteNamesFactory.GetPaletteNames(visualization);
```
Почему этот метод заполняет `availablePaletteNames`? Это необходимо?

balandin228 replied:
переменная выделена только для читаемости, класc ```PaletteNamesFactory``` просто получает названия всех палитр из класса ```CloudVIsualisation``` можно от него избавиться, и просто получить их через свойство 
```
availablePaletteNames = visualization.PaletteDictionary
                                            .Select(p => p.Key)
                                            .ToHashSet();
``` 
но я подумал, неплохо было бы сделать класс, который отвечал за получение этих названий палитр

balandin228 replied:
понял, в чем проблема, перенес эту отсветственность в конструктор

xackill commented on TagCloud/Client.cs:
```diff
@@ -25,123 +25,36 @@ public Client(IAction[] actions, IPaletteNamesFactory paletteNamesFactory)
         {
             availablePaletteNames = paletteNamesFactory.GetPaletteNames(visualization);
-            config = new ClientConfig(availableFontNames,availablePaletteNames);
+            config = new ClientConfig(availableFontNames, availablePaletteNames,visualization);
```
Чем идеологически различается `config` и `userSettings`? Может их удобнее объединить?

balandin228 replied:
``` config ``` отвечает за хранение информации о работе цикла программы.
1. Хранит картинку для вывода и сохранения
2. Отвечает за информацию о команде выхода 
3. Отвечает за информацию о том, что цикл запщуен
4 Отвечает за доступые именя шрифтов и палитр
то есть регулирует только работу цикла
``` userSettings ``` 
отвечает за все что касается пользователя и его настроек

balandin228 replied:
раньше все было вместе, но прошлый наставник как раз попросил разделить эти ответственности

xackill replied:
Звучит логично

xackill commented on TagCloud/Client.cs:
```diff
@@ -25,123 +25,36 @@ public Client(IAction[] actions, IPaletteNamesFactory paletteNamesFactory)
             availablePaletteNames = paletteNamesFactory.GetPaletteNames(visualization);
-            config = new ClientConfig(availableFontNames,availablePaletteNames);
+            config = new ClientConfig(availableFontNames, availablePaletteNames,visualization);
             var actionsDictionary = actions.ToDictionary(a => a.CommandName, a => a);
```
Почему каждый раз пересоздается `actionsDictionary`? Это необходимо?

balandin228 replied:
нет, не необходимо
done

xackill commented on TagCloud/Client.cs:
```diff
@@ -25,123 +25,36 @@ public Client(IAction[] actions, IPaletteNamesFactory paletteNamesFactory)
             var actionsDictionary = actions.ToDictionary(a => a.CommandName, a => a);
             Console.WriteLine("Список доступных комманд :");
             foreach (var action in actions)
                 Console.WriteLine($"{action.CommandName}     \"{action.Description}\"");
```
Зачем тут выводить список доступных команд, если есть отдельная команда для этого? Если хочется такую логику, то можно автоматически выполнить команду. Дублировать код — неудачное решение.

balandin228 replied:
done

xackill commented on TagCloud/Client.cs:
```diff
@@ -25,123 +25,36 @@ public Client(IAction[] actions, IPaletteNamesFactory paletteNamesFactory)
+                    var commandResult = ReadCommand(command, actionsDictionary);
+                    var perform = commandResult.Value.Perform(config, userSettings);
+                    if(!perform.IsSuccess)
+                        Console.WriteLine(perform.Error);
```
Та самая "гибридность", о которой я говорил выше.
Вроде как попытка поиспользовать паттерн `Result<T>`, но почему-то методы вызываются не в виде `fp`, а в чистом ооп. Предлагаю это порефакторить тело цикла так, чтобы поиспользовать паттерн `Result<T>` в явном виде.

- Заодно защититься от ошибки, когда ввели неверную команду.

balandin228 replied:
попробовал седлать более fp - шно, но проверку результата, как объекта, все равно придется оставить в цикле

xackill commented on TagCloud/Client.cs:
```diff
@@ -25,123 +25,36 @@ public Client(IAction[] actions, IPaletteNamesFactory paletteNamesFactory)
-            Console.WriteLine("Введенная палитра не поддерживается");
-            return false;
+            return availableCommands.ContainsKey(commandName)
+                ? Result.Ok(availableCommands[commandName])
```
Атата, двойной поиск по словарю. Подумай, как исправить на одиночный.

- Не единственное место в коде с такой проблемой. Необходимо исправить все!

balandin228 replied:
в этом исправил, пока ищу остальные места

xackill commented on TagCloud/Client.cs:
```diff
@@ -25,123 +25,36 @@ public Client(IAction[] actions, IPaletteNamesFactory paletteNamesFactory)
-        }
-
-        private bool TryReadPaletteName(out string paletteName)
+        private Result<IAction> ReadCommand(string commandName, Dictionary<string, IAction> availableCommands)
```
Название метода не актуально, теперь тут не происходит "чтение" команды.

balandin228 replied:
done

xackill commented on TagCloud/Models/FontSettings.cs:
```diff
@@ -11,17 +11,17 @@ public class FontSettings
             color = Color.Black;
         }
 
         public FontSettings(string fontName)
```
Сложность с этим классом в том, что `FontFamily` — `IDisposable` объект.
Стоит подумать о его времени жизни — кто и когда его будет освобождать.

balandin228 replied:
убрал работу с FontFamily
теперь работа только с FontFamilyName

xackill commented on TagCloud/Parsers/WordsHandler.cs:
```diff
@@ -24,12 +25,12 @@ public class WordsHandler : IWordsHandler
-                .ToDictionary(g => g.Key, g => g.Count()))
-                .ReplaceError(error => 
-                    error.Replace("Значение не может быть неопределенным.\r\n","Параметр не определен"));
+                    .SelectMany(s => Regex.Split(s.ToLower().Trim(), @"\W|_", RegexOptions.IgnoreCase))
```
При каждой итерации происходит компиляция `Regexp`-а, что точно не есть гуд для больших файлов.
Нужно понимать, в каких рамках это регулярное выражение изменяется и создавать его в том контексте, в рамках которого оно неизменяемо. В данном случае — это константа, которая не меняется в рамках всех экземпляров этого класса.

Ну и помнить, что создавая неизменяемое регулярное выражение, стоит его сразу скомпилировать.

balandin228 replied:
done

xackill commented on TagCloud/Parsers/WordsHandler.cs:
```diff
@@ -0,0 +1,43 @@
+
+namespace TagCloud
+{
+    public class WordsHandler : IWordsHandler
```
Убер-класс. В нем и:
1. Читается файл
2. Высчитывается кол-во слов в нем
3. Читаются скучные слова
4. Фильтруется по скучным словам

Проблема наличия таких классов — они не дают достаточной гибкости: нельзя безболезненно поменять поставщика слов (т.е. нельзя брать "скучные слова" не из файла BoringWords.txt, придется влезать в этот код и менять его. Стоит разбить этот класс на классы так, чтобы в каждом осталась ровно одна ответственность. Примеры отвественностей указаны сверху, к ним еще могу добавить такую:

5. Объединить ответственности 1-4 для получения результата.

balandin228 replied:
Как организовать эту гибкость? Допустим класс который отвечает за получение скучных слов? Допустим я разделю ответсвенность между 5ю классами, а в данный класс будет объединять эти ответственности. Вопрос, BoringWordsFarctory должен получать путь к файлу со скучными словами внутри класса WordsHandler где он будет вызываться. Или же BoringWordsFactory должен знать путь к файлу внутри себя, как статичный параметр?

xackill replied:
`WordsHandler` не должен зависеть от своих зависимостей: `IBoringWordsFarctory` должен лишь поставлять слова. Его реализация `FileBoringWords` зависит от пути к файлу. А его (гипотетическая) реализация `HttpsBoringWords` зависит от url и X509Certificate: при помощи сертификата устанавливается доверенное https-соединение с url и скачиваются слова.

Реализации разные, и их параметры не пересекаются. Классы-агрегаторы (WordsHandler) о внутренностях своих зависимостей ничего не знают; контейнер дал им `IBoringWordsFarctory`, с ним и работают.

По поводу статичности: настройки не должны быть статичными; они передаются в объект через контейнер. Т.е. при настройке контейнера указывается, что используется `FileBoringWords` в качестве `IBoringWordsFarctory` с параметром `FilePath = C:\boringWords.txt`

balandin228 replied:
done

xackill commented on TagCloud/Parsers/WordsHandler.cs:
```diff
@@ -0,0 +1,43 @@
+    public class WordsHandler : IWordsHandler
+    {
+        public Result<Dictionary<string, int>> Conversion(Dictionary<string, int> wordsAndCount)
+        {
```
Что касается самих подходов в текущих классах, то тут `fp` играет злую шутку: поскольку в одном методе содержится примерно все, то читаемость пропадает. Собственно необходимо это разбить по классам и в классах по методам так, чтобы используя паттерн `Result<T>` повысить читаемость.

xackill replied:
Можно использовать такой наглядный признак читаемости: в каждом из звеньев цепочки вызовов методов вызывается ровно один метод. Если необходимо какое-то преобразование чего-либо — это стоит выносить в отдельный метод.

#### Пример, где все действия объединены в одном (правда сложновато читать?)
```
Enumerable
    .Range(1, 5)
    .Select(i => string.Join("   ", Enumerable.Range(1, 5).Select(j => $"{i * j,2}")))
    .ToList()
    .ForEach(s => Console.WriteLine(s));
```

#### Пример с вынесением (тут кода побольше, зато каждая его часть читается проще)
```
const int numbersInLineCount = 5;

Enumerable
    .Range(1, numbersInLineCount)
    .Select(GetMultipliedNumbersLine) // method group
    .ToList()
    .ForEach(Console.WriteLine);

string GetMultipliedNumbersLine(int factor)
{
    var multipliedNumbers = Enumerable
        .Range(1, numbersInLineCount)
        .Select(i => $"{factor * i,2}");

    return string.Join("   ", multipliedNumbers);
}
```

#### О method group
Это хороший инструмент, который иногда сильно повышает читаемость, но следует помнить, что не всегда удается так удачно подобрать параметры и именования. Использование этой штуки не должно быть самоцелью, поэтому не следует его искусственно применять, от этого можно потерять в читаемости.

balandin228 replied:
в методе ```Conversion``` применил такой подход, читаться стало лучше.
в методе ```GetWordsAndCounts``` мне кажется будет лишним, так как создание коллекции наглядно видно в чейне из LINQ методов, а надстройка в виде Result.Of читаемость не сильно портит. Плюс еще цепочка не такая длинная, понять суть довольно легко

xackill commented on TagCloud/Result.cs:
```diff
@@ -0,0 +1,126 @@
+        }
+    }
+
+    public struct Result<T>
```
Почему `Result<T>` — это структура?
Какие это несет плюшки? А какие нюансы?

xackill commented on TagCloud/Actions/NewImageAction.cs:
```diff
@@ -8,15 +9,11 @@ public class NewImageAction : IAction
-            }
-
-            config.ToCreateNewImage = true;
+            Application.Exit();
```
Можешь рассказать, что именно делает эта операция?

balandin228 replied:
закрывает окно windows forms которое показывает картинку

xackill commented on TagCloud/Actions/SaveImageAction.cs:
```diff
@@ -13,46 +14,54 @@ public class SaveImageAction : IAction
+        private Result<ImageFormat> TryReadFormat()
         {
-            var defaultFormaName = "png";
+            var defaultFormatName = "png";
```
Дефолтные параметры находятся в отдельных `IAction`, чтобы их поменять, приходится искать конкретные места.
Намного проще объединить их в статический класс констант, а уже его использовать везде.
Будет всегда понятно, куда идти, если хочешь поменять дефолтные значения, а также сразу видна общая картина — какие дефолтные значения уже есть

balandin228 replied:
maybe done

xackill commented on TagCloud/Models/ClientConfig.cs:
```diff
@@ -5,8 +5,10 @@ namespace TagCloud
 {
     public class ClientConfig
```
У этой сущности есть нюанс: Bitmap, который она хранит внутри — IDisposable, который необходимо освобождать.

Как правило неконтролируемые CLR объекты (которые нужно освобождать вручную) долго не хранят — т.е. их оборачивают в `using` и они живут в рамках одного метода. Это удобно, потому что ты всегда знаешь, что у тебя нет утечки памяти из-за них. При текущей архитектуре это невозможно.

В рамках текущей задачи будет достаточно добавить корректное освобождение ресурса в рамках существующей архитектуры; однако на будущее стоит учитывать этот нюанс при проектировании архитектуры.

balandin228 replied:
сделал Bitmap.Dispose() в местах где создается новый Bitmap и приложение заканчивает работу

xackill commented on TagCloud/Models/ImageSettings.cs:
```diff
@@ -70,5 +67,28 @@ public void ReadPaletteName(string paletteName)
+            var settings = obj as ImageSettings;
+            return !(obj is null) &&
+                   Width == settings.Width && Height == settings.Height &&
+                   FontName == settings.FontName && PaletteName == settings.PaletteName;
```
Правда ли то условие?

balandin228 replied:
не понял, что с ним не так

xackill replied:
Разве не должно быть `!(settings is null)` ?
Иначе как обращаться к `settings.Width` дальше?

xackill commented on TagCloud/Factories/ImageFormatCollectionFactory.cs:
```diff
@@ -0,0 +1,21 @@
+            var namesFormatsToSave = new Dictionary<string, ImageFormat>();
+            namesFormatsToSave["jpg"] = ImageFormat.Jpeg;
+            namesFormatsToSave["png"] = ImageFormat.Png;
+            namesFormatsToSave["bmp"] = ImageFormat.Bmp;
```
Попробуй поиспользовать `Object initializer`, он проще в чтении такой инициализации :)

balandin228 replied:
done

xackill commented on TagCloudTests/ClientTests.cs:
```diff
@@ -0,0 +1,42 @@
+        private WindsorContainer container;
+
+        [Test]
+        public void IClient_Should_NotThrowsARuntimeExceptions_WhenStarts()
```
У этого теста много нюансов:
1. Он правда работает? Что будет, если вместо `client.Start()` написать `throw new Exception`?
2. Есть более удобные способы работать с потоками (через `Task`, но вам об этом еще расскажут)
3. `Thread.Sleep(5000);` — нативная реализация цикла `while`. Только не перепутать какой из потоков нужно уложить спать :)

balandin228 replied:
тест правда не работает
чтоб сделать так, чтоб он нормально работал, у меня не получилось
через Task пробовал, тоже не получилось

xackill replied:
На самом деле беда в `Assert` — проверяется не та сущность.
Необходимо проверять состояния тех потоков, которые запускаешь (через `Thread` или `Task`).

balandin228 committed:
add boringWordsFactory

перенес ответсвенность за чтение файла скучных слов в класс BoringWordsFactory
а ответсвенность за задание пути к фаул в класс TagCollectionFactory

balandin228 committed:
fix bugs

balandin228 committed:
add fileReader

вынес ответсвенность за чтение файла в класс fileReader

balandin228 committed:
fix  tagsCollection  responsibility

перенес чтение файла на уровень выше

balandin228 committed:
fix regex in fileReader

regex создается теперь один раз, при инициализации класса

balandin228 committed:
default settings

теперь во всех IAction используется статический метод класса UserSettings
GetDefaultSettings(), написал его давно, видимо забыл использовать

balandin228 committed:
убрал лишний код

balandin228 committed:
добавил object initializer

balandin228 committed:
fix tests

balandin228 committed:
Bitmap.Dispose()

добавил освобождение Bitmap при создании нового, и закрытии приложения

balandin228 committed:
Rename ReadCommand method

balandin228 committed:
поменял поле FontFamily на string

все работает со string

balandin228 committed:
try to fix ClientTest

balandin228 committed:
refactor Client

balandin228 committed:
try to use fp pattern

balandin228 committed:
убрал 2й обход

balandin228 committed:
вынес метод для читаемости

xackill commented:
#### Общий комментарий к коммитам 2
Стало значительно лучше — теперь можно посмотреть решение и валидировать каждый конкретный элемент. Однако есть еще немного шероховатостей:
1. Не в каждом коммите решение билдится; кое-где забыты тесты (например, в [cbfd1c4](https://github.com/kontur-courses/fp/pull/36/commits/cbfd1c478a2ab1f343b445a77fe84004d6574bd3))
2. Кое-где перемешаны коммиты (в [b91ce3d](https://github.com/kontur-courses/fp/pull/36/commits/b91ce3d28ef2bda115aa1bc540a4f9d90e0abd0e) есть части предыдущего решения)
3. Некоторые изменения слишком маленькие, их бы в какой-нибудь общий рефакторинг подключить (например [88c9786](https://github.com/kontur-courses/fp/pull/36/commits/88c97868da3fd1710756fddcf7a74100491fc6d7))

xackill commented:
#### Общий комментарий к решению 2
В целом сделан хороший шаг к использованию паттерна: проведена качественная работа с зависимостями, а основной цикл выглядит почти идеально :)
Единственное, в наследниках `IConsoleAction` паттерн снова не работает -- там нет цепочек, а отдельные вызовы — череваты исключениями.

Самое важное — порефакторить работу паттерна в наследниках `IConsoleAction`.

xackill commented on TagCloud/Client.cs:
```diff
@@ -31,22 +32,23 @@ public Client(IConsoleAction[] actions, IPaletteNamesFactory paletteNamesFactory
+                    .Then(command => GetCommandResult(command, actionsDictionary))
+                    .Then(commandResult => commandResult.Perform(config, userSettings));
+                if (!performResult.IsSuccess)
+                    Console.WriteLine(performResult.Error);
```
Насколько я понимаю, паттерн `Result<T>` позволяет обрабатывать этот случай. Найдешь в исходниках как?

xackill replied:
`GetCommandResult` сложно читать, как будто он отдает результат уже выполненной команды. А это не так.
Правда ли методе `GetCommandResult`  нужно применять паттерн `Result<T>`? `Then` обладает несколькими перегрузками.

balandin228 replied:
Поменял название, можно его сделать bool, а можно просто выкидывать Exception, ```Then```
обработает, но в другом контексте использования, например уже без ```Then```, мне показалось лучше вернуть Result<T>. На всякий случай

xackill commented on TagCloud/Factories/FileReader.cs:
```diff
@@ -11,10 +11,16 @@ namespace TagCloud.Factories
+
+        public FileReader()
+        {
+            regex = new Regex(@"\W|_", RegexOptions.IgnoreCase);
```
Дополнительно стоит помнить про компиляцию.
В таком виде `Regex` все равно при каждом использовании снова парсится из строки и превращается в конечный автомат.

balandin228 replied:
maybe Done

xackill commented on TagCloud/Models/DefaultSettings.cs:
```diff
@@ -0,0 +1,12 @@
+
+namespace TagCloud.Models
+{
+    public static class DefaultSettings
```
Так. Что-то неиспользуемое.

balandin228 replied:
убрал

xackill commented on TagCloud/Actions/GetPathConsoleAction.cs:
```diff
@@ -13,12 +13,13 @@ public class GetPathConsoleAction : IConsoleAction
 
         public Result<None> Perform(ClientConfig config, UserSettings settings)
         {
+            var defaultPathToRead = UserSettings.GetDefaultUserSettings().PathToRead;
```
При каждом вызове создаются новые и новые экземпляры, так точно не стоит делать — зачем генерировать поток объектов для GC?
Если беспокоит его "изменяемость", то предлагаю подумать, как это можно пофиксить :)

balandin228 replied:
сделал статическое поле в классе UserSettings

xackill commented on TagCloud/Parsers/WordsToTagsParser.cs:
```diff
@@ -27,7 +27,7 @@ public Result<List<Tag>> GetTags(Dictionary<string, int> words, ImageSettings im
         {
             var fontSize= fontSettings.defaultFontSize + count * 300 / words.Count ;
-            return new Font(fontSettings.fontFamily, (float) fontSize, fontSettings.fontStyle);
+            return new Font(fontSettings.fontFamilyName, (float) fontSize, fontSettings.fontStyle);
```
А работает все — потому что `Font` тоже `IDisposable`.
И его тоже надо освобождать.

balandin228 replied:
освободил в ```CloudVisualization```

xackill commented on TagCloud/Actions/ExitConsoleAction.cs:
```diff
@@ -12,6 +12,8 @@ public class ExitConsoleAction : IConsoleAction
         {
             config.ToExit = true;
+            if (!(config.ImageToSave is null))
+                config.ImageToSave.Dispose();
```
В C#6.0+ есть замечательный оператор `?.` :)

balandin228 replied:
перенес Dispose в другие ```IConsoleAction```

xackill commented on TagCloud/Actions/SaveImageConsoleAction.cs:
```diff
@@ -17,14 +17,17 @@ public SaveImageConsoleAction()
         public string Description { get; } = "save image";
 
         public Result<None> Perform(ClientConfig config, UserSettings settings)
         {
```
Давай на примере этого `IConsoleAction` рассмотрим, какие есть проблемы.

Для начала обращу внимание в принципе на читаемость — как известно, идеальный метод должен быть длиной в 3 строки. Конечно, соблюдать это правило получится далеко не всегда. Однако полэкрана текста не стоит оставлять: это нужно разбить по методам и, как предлагает задание, связать через `fp`. А вот теперь о нем и почему его так сложно сюда прикручивать.

Как правило, использование `fp` связывают с т.н. "чистыми функциями". Что такое "чистая функция"? Она читает пришедшие данные, на основе их создает новый объект и отдает его. Такая функция не меняет состояние пришедших на вход объектов и не меняет своего состояния (например, состояния класса, в котором она находится).
Получается, сейчас классы настроек совершенно для этого не подходят: у них другая логика.

Как это можно максимально прибрать?
Можно придумать разные решения, я предложу вот такое:
1. Все доп настройки (чтение путей сохранения, чтение форматов) — необходимо вынести отдельными `IConsoleAction`. Каждый `IConsoleAction` должен выполнять одну задачу (помним про принцип SRP).
2. В каждом таком `IConsoleAction` необходимо устроить чтение по типу "чистых функций", а последним действием (если все хорошо), сохранять в нужный класс настроек. Получается, что все действия смогут быть выполнены в необходимом паттерне.
3. Отказаться от хранения Bitmap в настройках. При сохранении (или отображении) создавать ее снова. После завершения использования — не забывать ее `Dispose`. С этим могут возникнуть проблемы, поскольку логика паттерна не предполагает обязательных вызовов Dispose, если какая-то из последующих частей развалилась. Вот тут стоит подумать, как лучше сделать. Может как раз сохранение/отображение картинки написать в процедурном стиле.

Собственно задача такая же: применить паттерн `Result<T>` в тех `IConsoleAction`, где это уместно. Сделаешь ли это по предложенному варианту или придумаешь что-нибудь другое — решать тебе :)

balandin228 replied:
done

balandin228 committed:
use Result.OnFail

balandin228 committed:
try to use Fp pattern in SaveImageAction

balandin228 committed:
Default Settings

- Поменял где нужно создание нового экзмепляра, на обращение к статическому полю класса UserSettings
- Но метод ```GetDefaultUserSetttings()``` решил оставить, так как, создание нового объекта настроек необходимо в классе ```Client```
Так же оставил метод изменения текущих настроек ```MakeDefault()``` в класса ```NewImageAction```.
Можно было бы без него, но мне кажется это лучше чем использовать ref

balandin228 committed:
Disposing Font

balandin228 committed:
add IFormatReader

balandin228 committed:
use fp pattern in SaveImageAction

balandin228 committed:
fp pattern in GetFontNameConsoleAction

balandin228 committed:
fp pattern in GetWidth and Height Actions

balandin228 committed:
refactor

balandin228 committed:
fp pattern in ShowImageConsoleAction + Disposing bitmap

balandin228 committed:
fp pattern in ReadPalette name Action

balandin228 committed:
refactor to Compile

balandin228 committed:
fp pattern in helpCommand and GetPath Actions

balandin228 committed:
add RegexOptions.Copiled

balandin228 committed:
use fp pattern in ExitConsoleAction

balandin228 committed:
code CleanUp

balandin228 committed:
fp in newImageConsoleAction

balandin228 committed:
more CleanUp

xackill commented on TagCloud/Actions/GetFontNameConsoleAction.cs:
```diff
@@ -14,14 +14,14 @@ public class GetFontNameConsoleAction : IConsoleAction
+            return ReadFontNameResult(config.AvailableFontNames)
+                .Then(font => settings.ImageSettings.FontName = font)
+                .Then(r => Result.Ok())
+                .OnFail(error => Result.Fail<None>(error));
```
Последние две строчки часто повторяются (в других IConsoleAction), хотя по факту особой логики не несут.
Почему бы не добавить метод `.ThenOfAction` ?

xackill commented on TagCloud/Actions/SaveImageConsoleAction.cs:
```diff
@@ -24,29 +26,35 @@ public SaveImageConsoleAction()
+            return TryReadFormat()
+                .OnFail(error => Result.Fail<None>(error))
+                .Then(formatResult => ReadPath(formatResult))
+                .Then(path => TryToSaveImage(path, config, settings));
```
Разве строчка с `.OnFail` тут нужна? Что она делает?

xackill commented on TagCloud/Actions/SaveImageConsoleAction.cs:
```diff
@@ -24,29 +26,35 @@ public SaveImageConsoleAction()
+                .OnFail(r => Result.Fail<None>(r))
+                .Then(bitmap =>
+                {
+                    bitmap.Save(path);
```
Формат файла куда-то подевался

balandin228 replied:
вроде его можно так же указать в конце пути в string

xackill replied:
Подожди, расширение и байтовая структура (формат файла) никак не связаны между собой.

В конце пути — это расширение файла: Windows так понимает, каким приложением его открыть. Байтовая структура — какие байты в каком порядке составляют картинку. Байтовая структура файла задается пропущенным параметром ImageFormat. Если почитать код класса Bitmap, то там сказано, что по умолчанию сохраняемая байтовая структура — `png`, причем не важно, что будет указано в расширении файла.

xackill commented on TagCloud/Services/CloudVisualization.cs:
```diff
@@ -39,8 +39,11 @@ private Bitmap GetImageOrThrow(ImageSettings imageSettings, string path, Palette
                 var rectangles = RectanglesCustomizer.GetRectanglesWithPalette(palette, tagRectanglesResult.Value);
                 foreach (var rectangle in rectangles)
+                {
                     graphics.DrawString(rectangle.Tag.Text, rectangle.Tag.Font, new SolidBrush(rectangle.Color),
```
Но `SolidBrush` тоже `IDisposable` :)
При проектировании приложения хорошо бы почитать про классы, которые планируешь использовать, и отталкиваясь от этого знания уже строить архитектуру.

xackill commented:
Зачтено 👍 👍
